### PR TITLE
Replace useSingleShard param for partitionKey

### PR DESCRIPTION
**Pull request by alarbada**
### Description

Stacked from #35 

As discussed internally with @lovromazgon (thanks for the tip!), the connector user can use the `partitionKey` parameter to group records into  shards, which would be the more appropiate way to group records.

I also add a little unit test to check whether the kinesis records are properly generated.

### CODE REVIEW

File: destination/destination.go [hunk 1]
```diff
@@ -111,47 +110,34 @@ func (d *Destination) Open(ctx context.Context) error {
...
-	// create the put records request
 	for _, rec := range records {
-		pKey := string(rec.Key.Bytes())
-		if len(pKey) > 256 {
-			pKey = pKey[:256]
+		partitionKey := d.config.PartitionKey
+		if partitionKey == "" {
+			partitionKey = string(rec.Key.Bytes())
+			if len(partitionKey) > 256 {
+				partitionKey = partitionKey[:256]
```
Comment (lovromazgon) [inline destination/destination.go:121]: I don't like assumptions 😄 Well, not ones that I can't influence as an end-user. And especially ones that could lead to unexpected behavior without even a warning.

If we are truncating the key, we should at least log a warning message, so the user is aware of it. Arguably, we could just not truncate it and let the request fail. The user can use processors to explicitly format records into the required format, this includes truncating the key.

File: destination/config.go [hunk 2]
```diff
@@ -21,10 +21,7 @@ type Config struct {
...
 
-	// UseSingleShard is a boolean that determines whether to add records in batches using KinesisClient.PutRecords
-	// (ordering not guaranteed, records written to multiple shards)
-	// or to add records to a single shard using KinesisClient.PutRecord
-	// (preserves ordering, records written to a single shard)
-	// Defaults to false to take advantage of batching performance
-	UseSingleShard bool `json:"useSingleShard" validate:"required" default:"true"`
+	// PartitionKey represents the kinesis partition key. Use this to group
+	// records into one or multiple shards.
+	PartitionKey string `json:"partitionKey"`
```
Comment (lovromazgon) [inline destination/config.go:26]: This only lets me specify a static partition key, which is not really useful, since whatever I specify here will make records land in the same shard.

What would be useful however is if I could specify a [Go template](https://pkg.go.dev/text/template) here which is evaluated on each record. Then I can specify whatever field I want, or even a combination of fields. By default it would contain `{{ .Key.Bytes }}`, which is the current behavior.

File: destination/destination.go (hunk 1 above)
Comment (alarbada) [reply destination/destination.go]: Hi @lovromazgon 

You're right, I've just added a warning log message when the partition key is truncated. However, I'm not so sure about having the user make a new processor just to setup the key.

It enforces a more complex setup. I'd argue back that if the default behaviour for partitionKeyTemplate is to print the record key back, it should should trim it (and we document this situation), just as a nice thing to do for the user. I don't think that the default connector configuration should enforce 2 decoupled pieces (processor + connector), looks more complex to me.

File: destination/destination.go [hunk 3]
```diff
@@ -17,16 +17,17 @@ package destination
 import (
 	"context"
 	"fmt"
+	"html/template"
```
Comment (lovromazgon) [inline destination/destination.go:20]:
```suggestion
	"text/template"
```

File: destination/destination.go (hunk 1 above)
Comment (lovromazgon) [reply destination/destination.go]: Fair enough, as long as there's a warning so the user knows what's happening it's fine 👍 Can we document this behavior in the readme?

File: destination/destination.go (hunk 3 above)
Comment (alarbada) [reply destination/destination.go:20]: Old habits die hard. Thanks!

File: destination/destination.go (hunk 1 above)
Comment (alarbada) [reply destination/destination.go]: Just documented it

File: destination/destination_test.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func TestPartitionKey(t *testing.T) {
+	t.Run("with partition key template defined", func(t *testing.T) {
+		ctx := context.Background()
+		is := is.New(t)
+		d := Destination{}
+		err := d.Configure(ctx, map[string]string{
+			"partitionKeyTemplate": `{{ printf "%s" .Position }}`,
+		})
```
Comment (lovromazgon) [inline destination/destination_test.go:86]: IIUC, this test would succeed even if this parameter would be ignored, because we are testing the default behavior. That's unfortunate and can hide bugs, can we please change the template so it produces an expectation that's different from the default? 🙏

Comment (alarbada) [reply destination/destination_test.go:86]: Ops, did not think about that, thanks!

File: destination/destination_test.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func TestPartitionKey(t *testing.T) {
+	t.Run("with partition key template defined", func(t *testing.T) {
+		ctx := context.Background()
+		is := is.New(t)
+		d := Destination{}
+		err := d.Configure(ctx, map[string]string{
+			"partitionKeyTemplate": `{{ printf "%s" .Key }}`,
+		})
```
Comment (lovromazgon) [inline destination/destination_test.go:86]: Lol, actually my [comment](https://github.com/conduitio-labs/conduit-connector-kinesis/pull/41#discussion_r1624293298) was wrong, previously the test was taking the data from the `.Position` field, which is _not_ the default. Actually now the test is testing the default (i.e. taking the partition key from`.Key`). I think what threw me off was the fact that the content said `test-position`.

TL;DR: we should revert it back so that the template does something different from the default behavior, like take the data from `.Position` instead of `.Key`.

PS: next time call me out if I comment something that doesn't make sense 😉

**Review by lovromazgon — APPROVED** [review]
Approving in advance, the one comment is a nitpick anyway 👍
